### PR TITLE
【OSCP】在 SecretFlow 拆分学习中实现 Direction-based scoring 攻击

### PR DESCRIPTION
**Pull request by Rainysponge**
**Type of change**
fixed: #1238
- [ ] Add new papers (Please tell us why you think this paper is awesome!)
- [ ] Fix the category of an existing paper/papers (Please tell us the reasons)
- [x] Add a new tool/primitive/application with a new markdown page (Thank you! Also, please tell us more about this awesome thing!)

**Description**

> This PR introduces the [direction-based scoring attack](https://arxiv.org/pdf/2102.08504.) into SecretFlow. The attack is a label inference attack for Vertical Federated Learning or Split Learning. By analyzing the direction of gradients, the attacker can infer labels by calculating the cosine similarity between the gradient of a sample with a known label and the gradients of other samples with unknown labels.

**Result**
We tested the code using the Criteo dataset and the Wide & Deep model, with settings similar to those described in the paper. The result is shown as follows:
image

### CODE REVIEW

File: secretflow/ml/nn/sl/attacks/direction_based_scoring_torch.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        multi_class = 'raise'
+        average = 'binary'
+        if self.num_classes > 2:
+            if len(label_targets.shape) == 1:
+                label_targets = label_targets.reshape(-1, 1)
+            if len(label_preds.shape) == 1:
+                label_preds = label_preds.reshape(-1, 1)
+            enc = OneHotEncoder(categories=[range(self.num_classes)], sparse=False)
+            label_targets = enc.fit_transform(label_targets)
+            label_preds = enc.fit_transform(label_preds)
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/attacks/direction_based_scoring_torch.py:166]: 如果是多分类，label_preds 也还是二分类 然后算metric？

File: tests/ml/nn/sl/attack/model_def.py [hunk 2]
```diff
@@ -259,3 +261,260 @@ def forward(self, x):
...
+            self.feature_index[feat] = start
+            start += 1
+
+        self.dnn = WideDeepBase(
+            len(self.dense_feature_columns) + embedding_size * len(self.embedding_dic),
+            dnn_hidden_units,
+            0.5,
+        )
+
+        self.dnn_linear = nn.Linear(dnn_hidden_units[-1], 1, bias=False)
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/attack/model_def.py:340]: 下面这些好像没用到

File: tests/ml/nn/sl/attack/model_def.py [hunk 3]
```diff
@@ -259,3 +261,260 @@ def forward(self, x):
...
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/attack/model_def.py:419]: 上面的好像没用到

File: tests/ml/nn/sl/attack/model_def.py [hunk 4]
```diff
@@ -259,3 +261,260 @@ def forward(self, x):
...
+        ]
+        sparse_input = torch.cat(sparse_embedding, dim=1)
+        sparse_input = torch.flatten(sparse_input, start_dim=1)
+        dense_values = [
+            X[:, self.feature_index[feat[0]]].reshape(-1, 1)
+            for feat in self.dense_feature_columns
+        ]
+        dense_input = torch.cat(dense_values, dim=1)
+        dnn_input = torch.cat((sparse_input, dense_input), dim=1)
+        dnn_out = self.dnn(dnn_input)
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/attack/model_def.py:373]: 这里是只有dense没有wide吗

File: tests/ml/nn/sl/attack/model_def.py [hunk 5]
```diff
@@ -259,3 +261,260 @@ def forward(self, x):
...
+        for feat in feat_size:
+            self.feature_index[feat] = start
+            start += 1
+
+        self.dnn = WideDeepBase(
+            len(self.dense_feature_columns) + embedding_size * len(self.embedding_dic),
+            dnn_hidden_units,
+            0.5,
+        )
+
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/attack/model_def.py:493]: 没用到的变量删一删

File: tests/ml/nn/sl/attack/test_torch_direct_based_scoring_lia.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    fixlen_feature_columns = [(feat, 'sparse') for feat in sparse_feature] + [
+        (feat, 'dense') for feat in dense_feature
+    ]
+    dnn_feature_columns = fixlen_feature_columns
+    linear_feature_columns = fixlen_feature_columns
+
+    # Split the Dataset as the settings in the paper
+    data = create_df(
+        source=df,
+        parts={
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/attack/test_torch_direct_based_scoring_lia.py:84]: 两方的特征是一样的？

File: tests/ml/nn/sl/attack/model_def.py (hunk 4 above)
Comment (zhaocaibei123) [reply tests/ml/nn/sl/attack/model_def.py:373]: 可以参考 secretflow/ml/nn/applications/sl_deepfm_torch.py base_model 可以输出多个 output

Comment (zhaocaibei123) [reply tests/ml/nn/sl/attack/model_def.py:351]: 如沟通，论文代码中是这个设定

**Review by zhaocaibei123 — APPROVED** [review]

File: secretflow/ml/nn/sl/attacks/direction_based_scoring_torch.py (hunk 1 above)
Comment (Rainysponge) [reply secretflow/ml/nn/sl/attacks/direction_based_scoring_torch.py]: 这个我现在已经删除相关代码 仅保留二分类了

File: tests/ml/nn/sl/attack/test_torch_direct_based_scoring_lia.py (hunk 6 above)
Comment (Rainysponge) [reply tests/ml/nn/sl/attack/test_torch_direct_based_scoring_lia.py:79]: 确实是一样的

**Review by zhaocaibei123 — APPROVED** [review]

**Review by ian-huu — APPROVED** [review]
LGTM
